boot/boot_robustness_test.go

```diff
@@ -0,0 +1,302 @@
+func runBootloaderLogic(c *C, bl bootloader.Bootloader) (snap.PlaceInfo, error) {
+	// switch on which kind of bootloader we have
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if ok {
+		return extractedRunKernelImageBootloaderLogic(c, ebl)
+	}
+
+	return pureenvBootloaderLogic(c, "kernel_status", bl)
+}
+
+// runBootloaderLogic implements the logic from the gadget snap bootloader,
+// namely that we transition kernel_status "try" -> "trying" and "trying" -> ""
+// and use try-kernel.efi when kernel_status is "try" and kernel.efi in all
+// other situations
+func extractedRunKernelImageBootloaderLogic(c *C, ebl bootloader.ExtractedRunKernelImageBootloader) (snap.PlaceInfo, error) {
+	m, err := ebl.GetBootVars("kernel_status")
+	c.Assert(err, IsNil)
+	kernStatus := m["kernel_status"]
+
+	kern, err := ebl.Kernel()
+	c.Assert(err, IsNil)
+	c.Assert(kern, Not(IsNil))
+
+	switch kernStatus {
+	case boot.DefaultStatus:
+	case boot.TryStatus:
+		// move to trying, use the try-kernel
+		m["kernel_status"] = boot.TryingStatus
+
+		// ensure that the try-kernel exists
+		tryKern, err := ebl.TryKernel()
+		c.Assert(err, IsNil)
+		c.Assert(tryKern, Not(IsNil))
+		kern = tryKern
+
+	case boot.TryingStatus:
+		// boot failed, move back to default
+		m["kernel_status"] = boot.DefaultStatus
+	}
+
+	err = ebl.SetBootVars(m)
+	c.Assert(err, IsNil)
+
+	return kern, nil
+}
+
+func pureenvBootloaderLogic(c *C, modeVar string, bl bootloader.Bootloader) (snap.PlaceInfo, error) {
+	m, err := bl.GetBootVars(modeVar, "snap_kernel", "snap_try_kernel")
+	c.Assert(err, IsNil)
+	var kern snap.PlaceInfo
+
+	kernStatus := m[modeVar]
+
+	kern, err = snap.ParsePlaceInfoFromSnapFileName(m["snap_kernel"])
+	c.Assert(err, IsNil)
+	c.Assert(kern, Not(IsNil))
+
+	switch kernStatus {
+	case boot.DefaultStatus:
+		// nothing to do, use normal kernel
+
+	case boot.TryStatus:
+		// move to trying, use the try-kernel
+		m[modeVar] = boot.TryingStatus
+
+		tryKern, err := snap.ParsePlaceInfoFromSnapFileName(m["snap_try_kernel"])
+		c.Assert(err, IsNil)
+		c.Assert(tryKern, Not(IsNil))
+		kern = tryKern
+
+	case boot.TryingStatus:
+		// boot failed, move back to default status
+		m[modeVar] = boot.DefaultStatus
+
+	}
+
+	err = bl.SetBootVars(m)
+	c.Assert(err, IsNil)
+
+	return kern, nil
+}
+
+// note: this could be implemented just as a function which takes a bootloader
+// as an argument and then inspect the type of MockBootloader that was passed
+// in, but the gains are little, since we don't need to use this function for
+// the non-ExtractedRunKernelImageBootloader implementations, as those
+// implementations just have one critical function to run which is just
+// SetBootVars
+func (s *bootenv20Suite) checkBootStateAfterUnexpectedRebootAndCleanup(
+	c *C,
+	dev boot.Device,
+	bootFunc func(boot.Device) error,
+	panicFunc string,
+	expectedBootedKernel snap.PlaceInfo,
+	expectedModeenvCurrentKernels []snap.PlaceInfo,
+	finalBlKernel snap.PlaceInfo,
```

Comment:
I don't see any test where finalBlKernel and expectedBootedKernel are different? do we expect some

Reply:
hmmmm yes I think you're right and we can just remove that arg too

Reply:
I changed this arg to be the expected kernel snap after the reboot, but before the final call to MarkSuccessful, and instead just assert that the final kernel on the bootloader is the same as the one the bootloader booted.